### DISCUSSION TITLE
Passwordless Login Link security versus password-based authentication

### DISCUSSION THREAD

**Discussion by kekkoudesu** [symfony/symfony#58239] [Q&A]
I was interested in implementing a passwordless login link, and it's cool that Symfony has this built in since 5.2: https://symfony.com/doc/current/security/login_link.html

But it does mention that it's less secure than password-based authentication:

> Login links are a convenient way of authenticating users, but it is also considered less secure than a traditional username and password form. It is not recommended to use login links in security critical applications.

I was wondering why, and I thought maybe it was because of the hash generation method.

With passwords, the user provides a secret set of characters, Symfony hashes it one-way and stores the hash, and now the user can only authenticate with a password that evaluates to that hash.

With passwordless login links, you define how the hash is generated with `signature_properties`. The documentation uses the User ID as an example of a variable you could use to generate the hash and says later on:

> hash
>    A hash of expires, user and any configured signature properties. Whenever these change, the hash changes and previous login links are invalidated. 

But the User ID (email) is essentially public knowledge; not a secret like a password, and an attacker would know the timestamp they requested a login at so can work out `expires`. So isn't it possible that an attacker could re-create the hash based on that information and guess the login link?

—is what I thought until I finally saw this section: https://symfony.com/doc/current/security/login_link.html#hashing-strategy

> Once this first hash is processed and encoded in Base64, a new one is created from the first hash value and the `kernel.secret` container parameter. This allows Symfony to sign this final hash, which is contained in the login URL. The final hash is also a Base64 encoded SHA-256 hash.

So it's a two-step hashing process. The first hash is generated with (essentially) publicly-known information, and the `APP_SECRET` variable stored in `.env` is used to generate the final hash.

But because `APP_SECRET` is stored in plain text on the server, that means anyone who gets access to this environment variable can now generate valid login links. Unlike passwords, where even the server doesn't know what the actual password is (only the hashes), compromising `APP_SECRET` is equivalent to compromising all user accounts on a site.

That's probably why the documentation mentions that it's less secure.

I think there are trade-offs that make login links a more attractive option from a security perspective, though.

- Most users do not create unique, secure passwords even when forced to comply with password rules.
- Compromise of any other site (except the user's email provider) will not lead to credential stuffing attacks on this site.

I'm thinking of implementing only login-link based authentication because I think there are actually security benefits for 99% of users compared to passwords—though maybe disabling it for Admins who I know will have unique secure passwords.

**Comment by DylanKas** [comment]
I don't really believe that a login-link based authentication would be safer than a more traditional approach.
The issue is not really about how a malicious user would generate fake login link though it could be. But a login-link is a single point of failure system. You get access to an user email, you get access to whatever service they may use. 

You think that it could be better for 99% of users, but those users are probably not using 2-FA for their emails. That would be really dangerous especially if multiple critical services also implement login-link authentication.

There are a few way to mitigate security risks as mentioned in the symfony docs and more, but if you just work as much on improving your basic login based authentication, then you would get something safer for both your users and your site.

We can imagine a lot of scenarios where it would be detrimental to use a magic-link instead.

With regular authentication, even if your server is breached, the attacker would only have access to password hashes, which would be quite hard to crack if using proper hashing algorithms. Whereas if your attacker only manage to get your `APP_SECRET` your whole authentication system is compromised.

**Reply by kekkoudesu** [reply]
You raise some good points! And it's good to know some more of the reasoning behind what's written in the documentation. Enforcing app-based 2FA is definitely something that could improve security in either authentication scenario, which I looked into after writing this.

> You think that it could be better for 99% of users, but those users are probably not using 2-FA for their emails. That would be really dangerous especially if multiple critical services also implement login-link authentication.

The primary reason I'm considering this is because access to a user's email is _already_ equivalent to account compromise on many services because many services use a "Reset Password" process that only demands the user have access to the email they used to sign up on the site. Login links are just the same, except they don't require you to set a password after you click the link.

If a user is unable to secure their email account, I don't think there's much we can do for them except enforce 2FA. Unless you get rid of the "Reset Password" link (which just doesn't seem practical, _so many_ users forget their password), I don't think there's a way to get rid of this vulnerability.

> With regular authentication, even if your server is breached, the attacker would only have access to password hashes, which would be quite hard to crack if using proper hashing algorithms.

It does concern me that `APP_SECRET` is a single point of failure. If there's some way to compromise the environment variable without otherwise compromising the server and gaining access to the database, all your users are vulnerable, just like that.

I have to wonder if implementing passkeys would be any more secure or whether it would necessarily only be based on the singular `APP_SECRET` variable again. And I guess TOTP would be too...

Defense-in-depth is important, so that's one reason to consider password-based authentication. Although just compromising `APP_SECRET` is enough even if using password-based auth in a Reset Password world...

**Edit:** Maybe it's possible to add a second factor to the Login Link implementation to improve security.

If you added another field to the User entity (say, `user_secret`) and generated a new random 32-character `secret` and updated the user entity immediately before you generated the Login Link every time (and added it to the `signature_properties` key), an attacker would need to compromise both the environment variable and the database to generate valid login links.
